test/studies/ml/lib/Tensor.chpl

```diff
@@ -0,0 +1,925 @@
+            M.reshapeDomain({0..#n,0..#m});
+            forall (i,j) in M.domain with (ref M, ref this) {
+                M.data[i,j] = this.data[j,i];
+            }
+            return M;
+        }
+
+        // Normalizes the tensor to have unit frobenius norm
+        proc normalize() {
+            const norm = sqrt(frobeniusNormPowTwo(this));
+            const data = this.data / norm;
+            return new Tensor(data);
+        }
+
+        // Retruns new tensor with provided domain
+        proc reshape(dom_) {
+            const dom = domainFromShape((...dom_.shape));
+            var t = new Tensor(dom.rank,eltType);
+            t.reshapeDomain(dom);
+            t.data = for (i,a) in zip(t.domain,this.data) do a;
+            return t;
+        }
+
+        // Returns new tensor with provided shape
+        proc reshape(shape: int ...?d) {
+            const dom = domainFromShape((...shape));
+            return this.reshape(dom);            
+        }
+
+        // Returns new tensor with rank 1
+        proc flatten() {
+            const size = this.data.domain.size;
+            return this.reshape({0..#size});
+        }
+        proc degen() {this.degen("");}
+        proc degen(s... ?k) {
+            for i in this.domain {
+                const x = this[i];
+                if AutoMath.isnan(x) {
+                    writeln(this,(...s));
+                    err("NaN in tensor.");
+                }
+                if AutoMath.isinf(x) {
+                    writeln(this,(...s));
+                    err("Inf in tensor.");
+                }
+            }
+        }
+
+        // Returns a tensor with argument function applied to each element
+        proc fmap(fn) {
+            var t = new Tensor(rank,eltType);
+            t.reshapeDomain(this.domain);
+            t.data = fn(this.data);
+            return t;
+        }
+        
+        // Prints the tensor (only really works for rank 1 and 2)
+        proc writeThis(fw: IO.fileWriter) throws {
+            fw.write("tensor(");
+            const shape = this.shape;
+            var first: bool = true;
+            for (x,i) in zip(data,0..) {
+                const idx = nbase(shape,i);
+                if idx[rank - 1] == 0 {
+                    if !first {
+                        fw.write("\n       ");
+                    }
+                    fw.write("[");
+                }
+                fw.writef("%{##.##########}",x);
+                
+                if idx[rank - 1] < shape[rank - 1] - 1 {
+                    if rank == 1 then
+                        fw.write("\n        ");
+                    else
+                        fw.write("  ");
+                } else {
+                    fw.write("]");
+                }
+                first = false;
+            }
+            fw.writeln(", shape=",this.shape,")");
+        }
+
+        // Serializer for tensor: rank,...shape,...data
+        proc write(fw: IO.fileWriter) throws {
```

Comment:
Same point as before about writing a `serialize` method s.t. you can take advantage of the serializers IO stuff more directly. Not really necessary, but it could clean up the code a little bit.